### PR TITLE
Re-implemented StateMachine

### LINKED ISSUE #34: State Machine for SAHT

### PR DESCRIPTION
**Pull request by tream2001**
Created app_statemachine.c and app_statemachine.h

Note: The StateMachine will need to be updated and changed as the handling functions for each of the states are created. This is meant to create a framework that other features can be coordinated with.

closes #34

### CODE REVIEW

**Comment by alanssitis** [comment]
@tream2001 should this be a draft? I see a lot of `TODO`'s

File: firmware-nucleo/smart-air-hockey-table/App/app_statemachine.c [hunk 1]
```diff
@@ -0,0 +1,77 @@
+#include "app_statemachine.h"
+
+static volatile GameState currGameState;
+// TODO: Make winScore modifiable by the player through the OLED menu, for now it is hard-coded to 7
+static volatile uint8_t winScore = 7; // Number of points required for a player to win
+static volatile uint8_t playerScores[NUM_PLAYERS];
```
Comment (alanssitis) [inline firmware-nucleo/smart-air-hockey-table/App/app_statemachine.c:6]: I don't think I like scores being in an array

**Review by alanssitis — COMMENTED** [review]
Would it be better if we had a global game state struct that we pass around? Something like:
```
struct GlobalGameState {
  GameState gameState;
  uint8_t playerScores[NUM_PLAYERS]; // Not sure if array is a good idea for this
  uint8_t winScore;
}
```

Interrupts will be able to just take this object around instead of having a bunch of loose variables around

File: firmware-nucleo/smart-air-hockey-table/App/Inc/app_statemachine.h [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+#define INC_APP_STATEMACHINE_H_
+
+#include <inttypes.h>
+
+#define NUM_PLAYERS 2
+#define PLAYER_A_IDX 0
+#define PLAYER_B_IDX 1
+
+/* TODO:
+ * May be able to combine states for each player (e.g. GameState_StartAnimationA & GameState_StartAnimationB)
```
[the remark this answers is not in the document]
Comment (tream2001) [reply firmware-nucleo/smart-air-hockey-table/App/Inc/app_statemachine.h]: Removed comment in commit 3afe1ea

File: firmware-nucleo/smart-air-hockey-table/App/Inc/app_statemachine.h [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+	GameState_StartGame,		// Set-up sequence for new game
+	GameState_StartAnimationA,	// Playing start animation for player A
+	GameState_StartAnimationB,	// Playing start animation for player B
+	GameState_RunGame,			// Game is in progress
+	GameState_ScoreAnimationA,	// Player A scored (not the winning shot)
+	GameState_ScoreAnimationB,	// Player B scored (not the winning shot)
+	GameState_WinAnimationA,	// Player A won the game (winning shot)
+	GameState_WinAnimationB,	// Player B won the game (winning shot)
+	GameState_Count				// State used for boundary checking
+} GameState;
```
[the remark this answers is not in the document]
Comment (tream2001) [reply firmware-nucleo/smart-air-hockey-table/App/Inc/app_statemachine.h:24]: Added in 76ac56f

File: firmware-nucleo/smart-air-hockey-table/App/app_statemachine.c [hunk 4]
```diff
@@ -0,0 +1,77 @@
+#include "app_statemachine.h"
+
+static volatile GameState currGameState;
```
[the remark this answers is not in the document]
Comment (tream2001) [reply firmware-nucleo/smart-air-hockey-table/App/app_statemachine.c]: Added in f6d1962

[the remark this answers is not in the document]
Comment (tream2001) [reply firmware-nucleo/smart-air-hockey-table/App/app_statemachine.c]: Although, wouldn't it make more sense to initialize as GameState_SetUp?

File: firmware-nucleo/smart-air-hockey-table/App/app_statemachine.c (hunk 1 above)
Comment (tream2001) [reply firmware-nucleo/smart-air-hockey-table/App/app_statemachine.c]: Switched from a "player index" to a "player ID" in b009e5d

**Comment by tream2001** [comment]
> Would it be better if we had a global game state struct that we pass around?
I like the idea of using a stuct to contain all of the relevant game information that could be passed around. (Inadvertently, this renders most of the previous commits irrelevant)

> Interrupts will be able to just take this object around instead of having a bunch of loose variables around
However, I am unsure what interrupts you are suggesting. Instead, I think it makes sense to pass the previously created struct to the handling functions by reference.  That way, the handling functions can modify the struct without needing a global variable.  What do you think of this?

**Comment by tream2001** [comment]
Also, I was previously imagining us merging in the framework then implementing each of the handling functions as different feature branches.  However, I like the idea of keeping this a draft-pr and adding those handling functions to this instead.

File: firmware-nucleo/smart-air-hockey-table/App/app_statemachine.c (hunk 4 above)
[the remark this answers is not in the document]
Comment (alanssitis) [reply firmware-nucleo/smart-air-hockey-table/App/app_statemachine.c]: We should only set it to "setup" when we actually start setup

**Comment by alanssitis** [comment]
I think a global variable is the best way to go about this. Passing a reference of a value to arbitrary interrupts like a button press does not seem fun to set up, so setting the state as a global state makes the most sense.

Also, I propose we also have a function that sets the current game state and force all api through this function if they wanna change game state.
```
void App_StateMachine_SetState(GameState);
```

This way, we can enforce strict transitions and find bugs. This way, we can also enforce that an error state stays in error. This will help us find weird bugs down the line.

Also, add the init function to the header.

On another note, i think `GameTick` should be another file, `superloop.c`, and leave `state_machine.c` state-machine-logic-only.

This means that you will need to add a getter:
```
GameState App_StateMachine_GetState();
```

**Comment by alanssitis** [comment]
Also, amma change gitignore cuz there're some files that shouldn't be there.

**Comment by tream2001** [comment]
Ben and I have been talking about the state machine in lab and it sounds like each of us have a similar but slightly different idea for the state machine.  So maybe we should put a hold on this till man lab this week

**Pull request closed without merging**